docs/guide.md

```diff
@@ -0,0 +1,575 @@
+   read and compile your Bolt file and upload the compiled JSON to your application Rules.
+
+## Data Validation
+
+The Firebase database is "schemaless" - which means that, unless you specify otherwise, any
+type or structure of data can be written anywhere in the database. By specifying a specific
+schema, you can catch coding errors early, and prevent malicious programs from writing data
+that you don't expect.
+
+Lets say your application wants to allow users to write messages to your database. Each message
+can be up to 140 characters and must by _signed_ by the user. In Bolt, you can express this
+using a `type` statement:
+
+_posts.bolt_
+```javascript
+// Allow anyone to read the list of Posts.
+path /posts {
+  read() = true;
+}
+
+// All individual Posts are writable by anyone.
+path /posts/$id is Post {
+  write() = true;
+}
+
+type Post {
+  validate() = this.message.length <= 140;
+
+  message: String,
+  from: String
+}
+```
+
+This database allows for a collection of `Posts` to be stored at the `/posts` path. Each one
+must have a unique ID key. Note that a path expression (after the `path` keyword) can contain a
+_wildcard_ component. This matches any string, and the value of the match is available to be
+used in expressions, if desired.
+
+For example, writing data at `/posts/123` will match the `path` statement when `$id` is equal
+to (the string) '123'.
+
+The Post type allows for exactly two string properties in each post (message and
+from). It also ensures that no message is longer than 140 characters.
+
+Bolt type statements can contain a `validate()` method (defined as `validate() = <expression>`,
+where the expression evaluates to `true` if the data is valid (can be saved to the
+database). When the expression evaluates to `false`, the attempt to write the data will return
+an error to the Firebase client and the database will be unmodified.
+
+To access properties of a type in an expression, use the `this` variable  (e.g. `this.message`).
+
+    $ firebase-bolt < posts.bolt
+
+```JSON
+{
+  "rules": {
+    "posts": {
+      ".read": "true",
+      "$id": {
+        ".validate": "newData.hasChildren(['message', 'from']) && newData.child('message').val().length <= 140",
+        "message": {
+          ".validate": "newData.isString()"
+        },
+        "from": {
+          ".validate": "newData.isString()"
+        },
+        "$other": {
+          ".validate": "false"
+        },
+        ".write": "true"
+      }
+    }
+  }
+}
+```
+
+Bolt supports the built-in datatypes of `String`, `Number`, `Boolean`, `Object`, `Any`, and
+`Null` (`Null` is useful for specifying optional properties):
+
+_sample.bolt_
+```javascript
+path / is Sample;
+
+type Sample {
+  name: String,
+  age: Number,
+  isMember: Boolean,
+
+  // The | type-operator allows this type to be an Object or undefined (null in Firebase).
+  attributes: Object | Null
```

Comment:
attributes is very generic, is there another more concrete property that sounds naturally optional? maidenName? spouse?